File: robot-gitee-framework1/service.go
package framework

import (
	"net/http"
	"strconv"
	"time"

	"github.com/opensourceways/community-robot-lib/interrupts"
)

type Service interface {
	RegisterIssueHandler(IssueHandler)
	RegisterPullRequestHandler(PullRequestHandler)
	RegisterPushEventHandler(PushEventHandler)
	RegisterNoteEventHandler(NoteEventHandler)

	Run(int, time.Duration)
}

func NewService() Service {
	return &service{}
}

type service struct {
	handlers
}

func (s *service) Run(port int, timeout time.Duration) {
	d := dispatcher{h: s.handlers}

	defer interrupts.WaitForGracefulShutdown()

	interrupts.OnInterrupt(func() {
		d.wait()
	})

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {})

	http.Handle("/gitee-hook", &d)

	httpServer := &http.Server{Addr: ":" + strconv.Itoa(port)}

	interrupts.ListenAndServe(httpServer, timeout)
}

File: robot-gitee-framework-python/robotgitee/framework/service.py
from robotgitee.framework.dispatcher import _Dispatcher
from robotgitee.framework.handlers import _Handlers


def run(robot: object, port: int, timeout: int) -> None:
    h = _Handlers()
    h.register_handler(robot)

    hs = h.get_handlers()
    if len(hs) == 0:
        raise Exception("it is not a robot")

    d = _Dispatcher(('', port), hs)
    d.run()
